jib-core/src/main/java/com/google/cloud/tools/jib/async/AsyncDependencies.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.async;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Builds a list of dependency {@link ListenableFuture}s to wait on before calling a {@link
+ * Callable}.
+ */
+public class AsyncDependencies {
+
+  /**
+   * Initialize with a {@link ListeningExecutorService}.
+   *
+   * @param listeningExecutorService the {@link ListeningExecutorService}
+   * @return a new {@link AsyncDependencies}
+   */
+  public static AsyncDependencies using(ListeningExecutorService listeningExecutorService) {
+    return new AsyncDependencies(listeningExecutorService);
+  }
+
+  private final ListeningExecutorService listeningExecutorService;
+
+  /** Stores the list of {@link ListenableFuture}s to wait on. */
+  private final List<ListenableFuture<?>> futures = new ArrayList<>();
+
+  private AsyncDependencies(ListeningExecutorService listeningExecutorService) {
+    this.listeningExecutorService = listeningExecutorService;
+  }
+
+  /**
+   * Adds the future of an {@link AsyncStep}.
+   *
+   * @param asyncStep the {@link AsyncStep}
+   * @return this
+   */
+  public AsyncDependencies addStep(AsyncStep<?> asyncStep) {
+    futures.add(asyncStep.getFuture());
+    return this;
+  }
+
+  /**
+   * Adds the future of an {@link AsyncStep} wrapped in an outer {@link AsyncStep}.
+   *
+   * @param asyncStepOfAsyncStep the outer {@link AsyncStep}
+   * @return this
+   * @throws ExecutionException if an exception occurs during execution of {@code
+   *     asyncStepOfAsyncStep}
+   */
+  public AsyncDependencies addStepOfStep(AsyncStep<? extends AsyncStep<?>> asyncStepOfAsyncStep)
+      throws ExecutionException {
+    return addStep(NonBlockingSteps.get(asyncStepOfAsyncStep));
```

Comment:
IMO, I think it is safer to not call `NonBlockingSteps.get()` in this helper class. Using it may simplify caller's code in a mechanical way, but I think assuming that the given outermost step must have been completed may become a bit danger when people start using this without knowing the details. As a async dependency handler, I'd prefer that whatever asyn step you throw to the handler, whether completed or now, it should work. After all, I think it'd be awkward put in the Javadoc of this method that "the given `asyncStepOfAsyncStep` must have been completed."

That said, you may lose the extra utility methods, but I think it might actually make this class easiler to understand. These additional methods seem like too case-specific with the assumption that the outer step is completed. Things like `asyncStepOfAsyncStepsOfAsyncStep` or `asyncStepOfAsyncStepOfAsyncStep` sound a bit confusing too. At least it seems better to not call `NonBlockingSteps.get()`.

Reply:
Agreed, will try to move `NonBlockingSteps.get()` out of this class.